lib/winapi/profiling.c

```diff
@@ -19,6 +23,73 @@ BOOL QueryPerformanceFrequency (LARGE_INTEGER *lpFrequency)
 {
     assert(lpFrequency != NULL);
 
-    lpFrequency->QuadPart = 733333333;
+    *lpFrequency = tsc_freq;
     return TRUE;
 }
+
+static uint64_t get_cpu_multiplier (void)
+{
+    // Read EBL_CR_POWERON
+    const uint64_t poweron_reg = __readmsr(0x2A);
+
+    // Chapter 2.22 of https://www.intel.com/content/dam/develop/external/us/en/documents/335592-sdm-vol-4.pdf
+    const uint64_t multiplier_pattern = (poweron_reg >> 22) & 0b101111;
+
+    // https://web.archive.org/web/20040407043044/http://www.intel.com:80/design/mobile/datashts/29834006.pdf
+    // clang-format off
+    switch (multiplier_pattern) {
```

Comment:
Should we have a default case? No impact to functionality but might causes warnings with -wall?